### ISSUE TITLE
QScript referencing nonexistent buckets

### ISSUE THREAD

**Issue by alissapajer** [quasar-analytics/quasar#3200]
```sql
select name from newTests where count(backends{_}) = 12
```
This is the integration test `unsupportedTests.test`. QScript is as of PR #3197. The query incorrectly returns zero results.

<details>
  <summary>QScript</summary>
  
```
QScript (Optimized):
Map
├─ Subset(Take)
│  ├─ Unreferenced
│  ├─ From
│  │  ╰─ Filter
│  │     ├─ EquiJoin
│  │     │  ├─ Unreferenced
│  │     │  ├─ Map
│  │     │  │  ├─ ShiftedRead(/newTests, IncludeId)
│  │     │  │  ╰─ ConcatMaps
│  │     │  │     ├─ MakeMap
│  │     │  │     │  ├─ Constant(Str(identities))
│  │     │  │     │  ╰─ MakeMap
│  │     │  │     │     ├─ Constant(Str(qsu1))
│  │     │  │     │     ╰─ ProjectIndex
│  │     │  │     │        ├─ ○
│  │     │  │     │        ╰─ Constant(Int(0))
│  │     │  │     ╰─ MakeMap
│  │     │  │        ├─ Constant(Str(value))
│  │     │  │        ╰─ Guard
│  │     │  │           ├─ ProjectIndex
│  │     │  │           │  ├─ ○
│  │     │  │           │  ╰─ Constant(Int(1))
│  │     │  │           ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │           ├─ ProjectIndex
│  │     │  │           │  ├─ ○
│  │     │  │           │  ╰─ Constant(Int(1))
│  │     │  │           ╰─ Undefined
│  │     │  ├─ Reduce
│  │     │  │  ├─ LeftShift
│  │     │  │  │  ├─ Map
│  │     │  │  │  │  ├─ ShiftedRead(/newTests, IncludeId)
│  │     │  │  │  │  ╰─ ConcatMaps
│  │     │  │  │  │     ├─ MakeMap
│  │     │  │  │  │     │  ├─ Constant(Str(identities))
│  │     │  │  │  │     │  ╰─ MakeMap
│  │     │  │  │  │     │     ├─ Constant(Str(qsu1))
│  │     │  │  │  │     │     ╰─ ProjectIndex
│  │     │  │  │  │     │        ├─ ○
│  │     │  │  │  │     │        ╰─ Constant(Int(0))
│  │     │  │  │  │     ╰─ MakeMap
│  │     │  │  │  │        ├─ Constant(Str(value))
│  │     │  │  │  │        ╰─ Guard
│  │     │  │  │  │           ├─ Guard
│  │     │  │  │  │           │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  ├─ ○
│  │     │  │  │  │           │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │  │  │           │  ├─ ProjectKey
│  │     │  │  │  │           │  │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  │  ├─ ○
│  │     │  │  │  │           │  │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  │  ╰─ Constant(Str(backends))
│  │     │  │  │  │           │  ╰─ Undefined
│  │     │  │  │  │           ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │  │  │           ├─ Guard
│  │     │  │  │  │           │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  ├─ ○
│  │     │  │  │  │           │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │  │  │           │  ├─ ProjectKey
│  │     │  │  │  │           │  │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  │  ├─ ○
│  │     │  │  │  │           │  │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  │  ╰─ Constant(Str(backends))
│  │     │  │  │  │           │  ╰─ Undefined
│  │     │  │  │  │           ╰─ Undefined
│  │     │  │  │  ├─ Struct
│  │     │  │  │  │  ╰─ ProjectKey
│  │     │  │  │  │     ├─ ○
│  │     │  │  │  │     ╰─ Constant(Str(value))
│  │     │  │  │  ├─ IdStatus
│  │     │  │  │  │  ╰─ IdStatus(ExcludeId)
│  │     │  │  │  ╰─ Repair
│  │     │  │  │     ╰─ ConcatMaps
│  │     │  │  │        ├─ MakeMap
│  │     │  │  │        │  ├─ Constant(Str(identities))
│  │     │  │  │        │  ╰─ ProjectKey
│  │     │  │  │        │     ├─ LeftSide
│  │     │  │  │        │     ╰─ Constant(Str(identities))
│  │     │  │  │        ╰─ MakeMap
│  │     │  │  │           ├─ Constant(Str(value))
│  │     │  │  │           ╰─ RightSide
│  │     │  │  ├─ Bucket
│  │     │  │  │  ╰─ List
│  │     │  │  │     ╰─ ProjectKey
│  │     │  │  │        ├─ ProjectKey
│  │     │  │  │        │  ├─ ○
│  │     │  │  │        │  ╰─ Constant(Str(identities))
│  │     │  │  │        ╰─ Constant(Str(qsu1))
│  │     │  │  ├─ Reducers
│  │     │  │  │  ╰─ Count
│  │     │  │  │     ╰─ ProjectKey
│  │     │  │  │        ├─ ○
│  │     │  │  │        ╰─ Constant(Str(value))
│  │     │  │  ╰─ Repair
│  │     │  │     ╰─ ConcatMaps
│  │     │  │        ├─ MakeMap
│  │     │  │        │  ├─ Constant(Str(identities))
│  │     │  │        │  ╰─ ConcatMaps
│  │     │  │        │     ├─ ConcatMaps
│  │     │  │        │     │  ├─ ConcatMaps
│  │     │  │        │     │  │  ├─ MakeMap
│  │     │  │        │     │  │  │  ├─ Constant(Str(qsu9_0))
│  │     │  │        │     │  │  │  ╰─ ReduceIndex(-\/(0))
│  │     │  │        │     │  │  ╰─ MakeMap
│  │     │  │        │     │  │     ├─ Constant(Str(qsu9_1))
│  │     │  │        │     │  │     ╰─ ReduceIndex(-\/(1))
│  │     │  │        │     │  ╰─ MakeMap
│  │     │  │        │     │     ├─ Constant(Str(qsu9_2))
│  │     │  │        │     │     ╰─ ReduceIndex(-\/(2))
│  │     │  │        │     ╰─ MakeMap
│  │     │  │        │        ├─ Constant(Str(qsu9_3))
│  │     │  │        │        ╰─ ReduceIndex(-\/(3))
│  │     │  │        ╰─ MakeMap
│  │     │  │           ├─ Constant(Str(value))
│  │     │  │           ╰─ Eq
│  │     │  │              ├─ ReduceIndex(\/-(0))
│  │     │  │              ╰─ Constant(Int(12))
│  │     │  ├─ List
│  │     │  │  ├─ tuple
│  │     │  │  │  ├─ ProjectKey
│  │     │  │  │  │  ├─ ProjectKey
│  │     │  │  │  │  │  ├─ ○
│  │     │  │  │  │  │  ╰─ Constant(Str(identities))
│  │     │  │  │  │  ╰─ Constant(Str(qsu1))
│  │     │  │  │  ╰─ ProjectKey
│  │     │  │  │     ├─ ProjectKey
│  │     │  │  │     │  ├─ ○
│  │     │  │  │     │  ╰─ Constant(Str(identities))
│  │     │  │  │     ╰─ Constant(Str(qsu9_0))
│  │     │  │  ├─ tuple
│  │     │  │  │  ├─ ProjectKey
│  │     │  │  │  │  ├─ ProjectKey
│  │     │  │  │  │  │  ├─ ○
│  │     │  │  │  │  │  ╰─ Constant(Str(identities))
│  │     │  │  │  │  ╰─ Constant(Str(qsu1))
│  │     │  │  │  ╰─ ProjectKey
│  │     │  │  │     ├─ ProjectKey
│  │     │  │  │     │  ├─ ○
│  │     │  │  │     │  ╰─ Constant(Str(identities))
│  │     │  │  │     ╰─ Constant(Str(qsu9_1))
│  │     │  │  ├─ tuple
│  │     │  │  │  ├─ ProjectKey
│  │     │  │  │  │  ├─ ProjectKey
│  │     │  │  │  │  │  ├─ ○
│  │     │  │  │  │  │  ╰─ Constant(Str(identities))
│  │     │  │  │  │  ╰─ Constant(Str(qsu1))
│  │     │  │  │  ╰─ ProjectKey
│  │     │  │  │     ├─ ProjectKey
│  │     │  │  │     │  ├─ ○
│  │     │  │  │     │  ╰─ Constant(Str(identities))
│  │     │  │  │     ╰─ Constant(Str(qsu9_2))
│  │     │  │  ╰─ tuple
│  │     │  │     ├─ ProjectKey
│  │     │  │     │  ├─ ProjectKey
│  │     │  │     │  │  ├─ ○
│  │     │  │     │  │  ╰─ Constant(Str(identities))
│  │     │  │     │  ╰─ Constant(Str(qsu1))
│  │     │  │     ╰─ ProjectKey
│  │     │  │        ├─ ProjectKey
│  │     │  │        │  ├─ ○
│  │     │  │        │  ╰─ Constant(Str(identities))
│  │     │  │        ╰─ Constant(Str(qsu9_3))
│  │     │  ├─ Inner
│  │     │  ╰─ ConcatArrays
│  │     │     ├─ MakeArray
│  │     │     │  ╰─ LeftSide
│  │     │     ╰─ MakeArray
│  │     │        ╰─ RightSide
│  │     ╰─ ProjectKey
│  │        ├─ ProjectIndex
│  │        │  ├─ ○
│  │        │  ╰─ Constant(Int(1))
│  │        ╰─ Constant(Str(value))
│  ╰─ Count
│     ╰─ Map
│        ├─ Unreferenced
│        ╰─ Constant(Int(11))
╰─ ProjectKey
   ├─ ProjectKey
   │  ├─ ProjectIndex
   │  │  ├─ ○
   │  │  ╰─ Constant(Int(0))
   │  ╰─ Constant(Str(value))
   ╰─ Constant(Str(name))
```
</details>

**Comment by djspiewak** [comment]
@alissapajer Is this actually qsu?  I thought it was the equijoin rewrite that was doing this.

**Comment by alissapajer** [comment]
@djspiewak this is the optimized qscript after it's been through qsu. Does that answer your question?

**Comment by alissapajer** [comment]
Optimized for `mimir`, if I remember correctly. Should have taken note of that.

**Comment by djspiewak** [comment]
Right but qsu isn't where the invalid array access gets generated, I thought.

**Comment by wemrysi** [comment]
@djspiewak It's not invalid array access, it's non-existent bucket references in join conditions, due to duplicate provenance.

**Comment by alissapajer** [comment]
Given the fixed non-existent bucket access on @djspiewak 's branch, there was talk of sparse arrays causing this to fail in mimir. Those sparse arrays would be created during connector-specific rewrites after qsu.

**Comment by djspiewak** [comment]
Ah I see.   Yeah as of my PR, it's just blocked on the sparse arrays.

**Comment by djspiewak** [comment]
@alissapajer Now that this test is passing, should I mark the PR as fixing it?

**Comment by alissapajer** [comment]
@djspiewak do we know why this is failing for mongo as of https://github.com/quasar-analytics/quasar/pull/3212/files#diff-88ecb580ddac0385ae52ab3bfd380735R10? Is latest master 805d22d (includes some mongo fixes) merged into that PR?

**Comment by djspiewak** [comment]
@alissapajer Latest master isn't merged. I believe there were _oid issues, so latest master probably fixes it.  I'll make another PR that looks into that and closes this one if so.

**Comment by djspiewak** [comment]
@alissapajer @wemrysi Merging master into `bug/over-aggressive-coalesce` (which is to say, master) and trying this query on both mongo and mimir, I get no results whatsoever.  That seems like a weird regression:

```
QScript (Optimized):
Map
├─ Subset(Take)
│  ├─ Unreferenced
│  ├─ From
│  │  ╰─ Filter
│  │     ├─ EquiJoin
│  │     │  ├─ Unreferenced
│  │     │  ├─ Map
│  │     │  │  ├─ ShiftedRead(/newTests, IncludeId)
│  │     │  │  ╰─ ConcatMaps
│  │     │  │     ├─ MakeMap
│  │     │  │     │  ├─ Constant(Str(identities))
│  │     │  │     │  ╰─ MakeMap
│  │     │  │     │     ├─ Constant(Str(qsu1))
│  │     │  │     │     ╰─ ProjectIndex
│  │     │  │     │        ├─ ○
│  │     │  │     │        ╰─ Constant(Int(0))
│  │     │  │     ╰─ MakeMap
│  │     │  │        ├─ Constant(Str(value))
│  │     │  │        ╰─ Guard
│  │     │  │           ├─ ProjectIndex
│  │     │  │           │  ├─ ○
│  │     │  │           │  ╰─ Constant(Int(1))
│  │     │  │           ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │           ├─ ProjectIndex
│  │     │  │           │  ├─ ○
│  │     │  │           │  ╰─ Constant(Int(1))
│  │     │  │           ╰─ Undefined
│  │     │  ├─ Reduce
│  │     │  │  ├─ LeftShift
│  │     │  │  │  ├─ Map
│  │     │  │  │  │  ├─ ShiftedRead(/newTests, IncludeId)
│  │     │  │  │  │  ╰─ ConcatMaps
│  │     │  │  │  │     ├─ MakeMap
│  │     │  │  │  │     │  ├─ Constant(Str(identities))
│  │     │  │  │  │     │  ╰─ MakeMap
│  │     │  │  │  │     │     ├─ Constant(Str(qsu1))
│  │     │  │  │  │     │     ╰─ ProjectIndex
│  │     │  │  │  │     │        ├─ ○
│  │     │  │  │  │     │        ╰─ Constant(Int(0))
│  │     │  │  │  │     ╰─ MakeMap
│  │     │  │  │  │        ├─ Constant(Str(value))
│  │     │  │  │  │        ╰─ Guard
│  │     │  │  │  │           ├─ Guard
│  │     │  │  │  │           │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  ├─ ○
│  │     │  │  │  │           │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │  │  │           │  ├─ ProjectKey
│  │     │  │  │  │           │  │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  │  ├─ ○
│  │     │  │  │  │           │  │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  │  ╰─ Constant(Str(backends))
│  │     │  │  │  │           │  ╰─ Undefined
│  │     │  │  │  │           ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │  │  │           ├─ Guard
│  │     │  │  │  │           │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  ├─ ○
│  │     │  │  │  │           │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  ├─ Type(Obj(Map[], Some(Top)))
│  │     │  │  │  │           │  ├─ ProjectKey
│  │     │  │  │  │           │  │  ├─ ProjectIndex
│  │     │  │  │  │           │  │  │  ├─ ○
│  │     │  │  │  │           │  │  │  ╰─ Constant(Int(1))
│  │     │  │  │  │           │  │  ╰─ Constant(Str(backends))
│  │     │  │  │  │           │  ╰─ Undefined
│  │     │  │  │  │           ╰─ Undefined
│  │     │  │  │  ├─ Struct
│  │     │  │  │  │  ╰─ ProjectKey
│  │     │  │  │  │     ├─ ○
│  │     │  │  │  │     ╰─ Constant(Str(value))
│  │     │  │  │  ├─ IdStatus
│  │     │  │  │  │  ╰─ IdStatus(ExcludeId)
│  │     │  │  │  ╰─ Repair
│  │     │  │  │     ╰─ ConcatMaps
│  │     │  │  │        ├─ MakeMap
│  │     │  │  │        │  ├─ Constant(Str(identities))
│  │     │  │  │        │  ╰─ ProjectKey
│  │     │  │  │        │     ├─ LeftSide
│  │     │  │  │        │     ╰─ Constant(Str(identities))
│  │     │  │  │        ╰─ MakeMap
│  │     │  │  │           ├─ Constant(Str(value))
│  │     │  │  │           ╰─ RightSide
│  │     │  │  ├─ Bucket
│  │     │  │  │  ╰─ List
│  │     │  │  │     ╰─ ProjectKey
│  │     │  │  │        ├─ ProjectKey
│  │     │  │  │        │  ├─ ○
│  │     │  │  │        │  ╰─ Constant(Str(identities))
│  │     │  │  │        ╰─ Constant(Str(qsu1))
│  │     │  │  ├─ Reducers
│  │     │  │  │  ╰─ Count
│  │     │  │  │     ╰─ ProjectKey
│  │     │  │  │        ├─ ○
│  │     │  │  │        ╰─ Constant(Str(value))
│  │     │  │  ╰─ Repair
│  │     │  │     ╰─ ConcatMaps
│  │     │  │        ├─ MakeMap
│  │     │  │        │  ├─ Constant(Str(identities))
│  │     │  │        │  ╰─ MakeMap
│  │     │  │        │     ├─ Constant(Str(qsu9_0))
│  │     │  │        │     ╰─ ReduceIndex(-\/(0))
│  │     │  │        ╰─ MakeMap
│  │     │  │           ├─ Constant(Str(value))
│  │     │  │           ╰─ Eq
│  │     │  │              ├─ ReduceIndex(\/-(0))
│  │     │  │              ╰─ Constant(Int(2))
│  │     │  ├─ List
│  │     │  │  ╰─ tuple
│  │     │  │     ├─ ProjectKey
│  │     │  │     │  ├─ ProjectKey
│  │     │  │     │  │  ├─ ○
│  │     │  │     │  │  ╰─ Constant(Str(identities))
│  │     │  │     │  ╰─ Constant(Str(qsu1))
│  │     │  │     ╰─ ProjectKey
│  │     │  │        ├─ ProjectKey
│  │     │  │        │  ├─ ○
│  │     │  │        │  ╰─ Constant(Str(identities))
│  │     │  │        ╰─ Constant(Str(qsu9_0))
│  │     │  ├─ Inner
│  │     │  ╰─ ConcatMaps
│  │     │     ├─ MakeMap
│  │     │     │  ├─ Constant(Str(left))
│  │     │     │  ╰─ LeftSide
│  │     │     ╰─ MakeMap
│  │     │        ├─ Constant(Str(right))
│  │     │        ╰─ RightSide
│  │     ╰─ ProjectKey
│  │        ├─ ProjectKey
│  │        │  ├─ ○
│  │        │  ╰─ Constant(Str(right))
│  │        ╰─ Constant(Str(value))
│  ╰─ Count
│     ╰─ Map
│        ├─ Unreferenced
│        ╰─ Constant(Int(11))
╰─ ProjectKey
   ├─ ProjectKey
   │  ├─ ProjectKey
   │  │  ├─ ○
   │  │  ╰─ Constant(Str(left))
   │  ╰─ Constant(Str(value))
   ╰─ Constant(Str(name))

Query time: 0.7s
No results found
```

**Comment by alissapajer** [comment]
@djspiewak what's the original query?

**Comment by alissapajer** [comment]
Oh, never mind, it's the query from the PR description :)

**Comment by wemrysi** [comment]
@djspiewak Looks like that is QScript for `select name from newTests where count(backends{_}) = 2`, not sure if that makes a difference or not. It looks correct to me.

**Comment by djspiewak** [comment]
Oh right, I changed the query in the repl.

**Comment by wemrysi** [comment]
@djspiewak That's why I was asking about the query, _are_ there any with 2 backends?

**Comment by djspiewak** [comment]
Yep, quite a few.

**Comment by wemrysi** [comment]
Then I'm not sure what's going on.

**Comment by djspiewak** [comment]
@wemrysi and I figured this out: it's just not working in mimir because of #3183.  This particular issue is done.
